pytential/symbolic/pde/reduce_fmms.py

```diff
@@ -0,0 +1,405 @@
+                if source_expr not in source_exprs:
+                    source_exprs.append(source_expr)
+                    row += [0]
+                poly = _kernel_source_derivs_as_poly(source_kernel, axis_vars)
+                row[source_exprs.index(source_expr)] += poly * to_sympy(coeff)
+        matrix.append(row)
+
+    # At the beginning, we didn't know the number of columns of the matrix.
+    # Therefore we used a list for rows and they kept expanding.
+    # Here we are adding zero padding to make the result look like a matrix.
+    for row in matrix:
+        row += [0]*(len(source_exprs) - len(row))
+
+    return matrix, source_exprs
+
+
+def _check_int_gs_common(int_gs):
+    """Checks that the :class:`~pytential.symbolic.primtive.IntG` objects
+    have the same base kernel and other properties that would allow
+    merging them.
+    """
+    base_kernel = int_gs[0].source_kernels[0].get_base_kernel()
+    common_int_g = int_gs[0].copy(target_kernel=base_kernel,
+            source_kernels=(base_kernel,), densities=(1,))
+    for int_g in int_gs:
+        for source_kernel in int_g.source_kernels:
+            if source_kernel.get_base_kernel() != base_kernel:
+                return False
+        if common_int_g != int_g.copy(target_kernel=base_kernel,
+                source_kernels=(base_kernel,), densities=(1,)):
+            return False
+    return True
+
+
+def _kernel_source_derivs_as_poly(kernel, axis_vars):
+    """Converts a :class:`sumpy.kernel.Kernel` object to a polynomial.
+    A :class:`sumpy.kernel.Kernel` represents a derivative operator
+    and the derivative operator is converted to a polynomial with
+    variables given by `axis_vars`.
+
+    For eg: for source x the derivative operator,
+    d/dx_1 dx_2 + d/dx_1 is converted to x_2 * x_1 + x_1.
+    """
+    if isinstance(kernel, AxisSourceDerivative):
+        poly = _kernel_source_derivs_as_poly(kernel.inner_kernel, axis_vars)
+        return -axis_vars[kernel.axis]*poly
+    if isinstance(kernel, KernelWrapper):
+        raise ValueError
+    return 1
+
+
+def _convert_source_poly_to_int_g(poly, orig_int_g, axis_vars):
+    """This does the opposite of :func:`_kernel_source_derivs_as_poly`
+    and converts a polynomial back to a source derivative
+    operator. First it is converted to a :class:`sumpy.kernel.Kernel`
+    and then to a :class:`~pytential.symbolic.primitives.IntG`.
+    """
+    from pytential.symbolic.pde.system_utils import simplify_densities
+    to_pymbolic = SympyToPymbolicMapper()
+
+    orig_kernel = orig_int_g.source_kernels[0]
+    source_kernels = []
+    densities = []
+    for monom, coeff in poly.terms():
+        kernel = orig_kernel
+        for idim, rep in enumerate(monom):
+            for _ in range(rep):
+                kernel = AxisSourceDerivative(idim, kernel)
+        source_kernels.append(kernel)
+        densities.append(to_pymbolic(coeff) * (-1)**sum(monom))
```

Comment:
Maybe comment on the reasoning for (-1)**sum(monom)

Reply:
Done in 2d659b08251fb864ab47da9d71bb7bef126092bb